### PR TITLE
Fix usage of addOrderingRules in SolutionArray

### LINKED ISSUE #1776: Ordering of SolutionArray components needs to be revisited.
**Problem description**

#1775 introduced a potential memory leak via incorrect use of `AnyMap::addOrderingRules`, see https://github.com/Cantera/cantera/pull/1775#discussion_r1718528240

### PR DESCRIPTION
**Pull request by speth**
**Changes proposed in this pull request**

- Fix order that "head" ordering rules are interpreted, so the first ones given have highest priority
- Fix extra quotes around `git-commit` field
- Avoid setting ordering rules for `SolutionArray` more than once

**If applicable, fill in the issue number this pull request is fixing**

Closes #1776

**If applicable, provide an example illustrating new features this pull request is introducing**

**Checklist**

- [x] The pull request includes a clear description of this code change
- [x] Commit messages have short titles and reference relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [x] The pull request is ready for review

### CODE REVIEW

File: src/base/SolutionArray.cpp [hunk 1]
```diff
@@ -924,9 +924,7 @@ AnyMap preamble(const string& desc)
     }
     data["generator"] = "Cantera SolutionArray";
     data["cantera-version"] = CANTERA_VERSION;
-    // escape commit to ensure commits are read correctly from YAML
-    // example: prevent '3491027e7' from being converted to an integer
-    data["git-commit"] = "'" + gitCommit() + "'";
+    data["git-commit"] = gitCommit();
```
Comment (ischoegl) [inline src/base/SolutionArray.cpp:927]: I added the quotes after running into an issue with a git hash being a number instead of a string (see comment).

Comment (ischoegl) [reply src/base/SolutionArray.cpp:927]: Nevermind: saw comment about #1740

**Review by ischoegl — CHANGES_REQUESTED** [review]
@speth - thanks for a simpler approach; you definitely have a better understanding of the inner workings of `AnyMap`! Unfortunately, this PR reverts my earlier fix to the previous situation, where overwriting an existing `SolutionArray` scrambles the order of entries. Looking at the output of the CI run `test-python-composite`, the created output `solutionarray_overwrite.yaml` is:
```YAML
arr:
  generator: Cantera SolutionArray
  cantera-version: 3.1.0a4
  git-commit: 527ff6ee4
  date: Mon Aug 19 15:54:54 2024
  data:
    size: 8
    basis: mass
    components: [T, D, H2, H, O, O2, OH, H2O, HO2, H2O2, AR, N2]
    D: [0.1494213406287969, 0.1477867702311559, 0.1461697603124980,
    0.1445732999152547, 0.1429960611432598, 0.1414355577644473,
    0.1398908449533592, 0.1383648238149952]
    T: [3070.786192328174, 3087.932654812, 3104.811774920454, 3121.397172300316,
    3137.707859111618, 3153.774699944349, 3169.612514141550,
    3185.196897445414]
[...]
```
with incorrect ordering - `T` should be listed first, but for whatever reason `D` appears on top.

**Comment by speth** [comment]
> Unfortunately, this PR reverts my earlier fix to the previous situation, where overwriting an existing `SolutionArray` scrambles the order of entries.

Not quite. The order of output is now different than what it was before, but it is now stable when overwriting the file. I've updated the corresponding test to check that this is indeed the case.

For a 1D flame, the output now comes out in the following order:
```yaml
  flame:
    type: free-flow
    size: 9
    basis: mass
    components: [grid, velocity, T, D, H2, H, O, O2, OH, H2O, HO2, H2O2, AR, N2]
    points: 9
    energy-enabled: true
    phase:
      name: ohmech
      source: /Users/speth/src/cantera/build/python/cantera/data/h2o2.yaml
    radiation-enabled: false
    species-enabled: true
    transport-model: mixture-averaged
    Soret-enabled: false
    flux-gradient-basis: 1
    refine-criteria:
      ratio: 3.0
      slope: 0.06
      curve: 0.12
      prune: 0.0
      grid-min: 1.0e-10
      max-points: 1000
    tolerances:
      transient-abstol: 1.0e-11
      steady-abstol: 1.0e-09
      transient-reltol: 1.0e-04
      steady-reltol: 1.0e-04
    fixed-point:
      location: 0.0105
      temperature: 698.1678485027377
    grid: [0.0, 6.0e-03, 9.0e-03, 0.0105, 0.012, 0.015, 0.018, 0.024, 0.03]
    velocity: [0.3448108359476526, 0.3465608656848906, 0.3770754165767459,
    0.5740023649233583, 1.364440990996746, 1.439569578030446,
    1.450116128788632, 1.452935503745143, 1.452935503745178]
    D: [1.004145957804174, 0.9990718314487402, 0.9182237865090841,
    0.6032044854535895, 0.2537620765431486, 0.2405193521271233,
    0.2387706120988350, 0.2383081252938835, 0.2383081252938287]
    T: [400.0, 403.1281230365552, 445.1340622371536, 698.1678485027377,
    1717.667237539325, 1819.506593089243, 1833.946369770278,
    1837.803402275211, 1837.803402275211]
    H2: [9.466124468899828e-03, 9.282907545332970e-03, 8.219365106913516e-03,
    5.573672033086918e-03, 1.659105275821406e-04, 2.220294855777577e-05,
    6.143983384802635e-06, 2.911456291310724e-06, 2.911456291310724e-06]
    H: [1.179218242823033e-15, 7.789514857054084e-15, 2.921328219627074e-11,
    1.809466438706173e-07, 4.695488413457595e-05, 2.476002464123360e-06,
    3.627949117616036e-07, 1.210646613146704e-07, 1.210646613146704e-07]
    ...
```

This matches the order that things are added to the output `AnyMap`: metadata first, then "extra" fields which include the grid, velocity, and any solution components that aren't part of the thermodynamic state, followed by the thermodynamic state variables, then with a few specific fields moved to the beginning. The order that `T`, `D`, and `Y` come out is dictated by the order provided by `phase->nativeState()`, which uses `std::map` and therefore returns items in alphabetical order. If you want them in the order used by `ThermoPhase`, then I guess we could use `nativeMode` instead?

**Comment by ischoegl** [comment]
> The order that `T`, `D`, and `Y` come out is dictated by the order provided by `phase->nativeState()`, which uses `std::map` and therefore returns items in alphabetical order. If you want them in the order used by `ThermoPhase`, then I guess we could use `nativeMode` instead?

Thanks for reminding me of this difference. Yes, things probably should be added in the same order as they show up in `components`, i.e. 
```
[grid, velocity, T, D, H2, H, O, O2, OH, H2O, HO2, H2O2, AR, N2]
```
which itself can be added at a later point so it immediately precedes the `grid` entry?

Beyond, we can probably banish `phase` to the tail? I honestly was too focused on making HDF5 behave when I implemented this and didn't pick up on those subtle differences.

**Comment by speth** [comment]
> Thanks for reminding me of this difference. Yes, things probably should be added in the same order as they show up in `components`, i.e.
> 
> ```
> [grid, velocity, T, D, H2, H, O, O2, OH, H2O, HO2, H2O2, AR, N2]
> ```

I've pushed an update that implements this, with `components` showing up just before `grid`.

> Beyond, we can probably banish `phase` to the tail? I honestly was too focused on making HDF5 behave when I implemented this and didn't pick up on those subtle differences.

I don't think we need to be too fussy about the formatting in this file. That said, I kind of like having the array data come last, as it can be quite large if you have a lot of grid points and species. The other settings for the 1D domain are much smaller by comparison, so it's kind of nice to see them up front.

**Review by ischoegl — DISMISSED** [review]
Thanks for fixing that part, @speth. We're still not quite there as the order of the leading extra's isn't quite there, but I'm willing to let this rest.

```
solution:
  date: Tue Aug 20 21:25:14 2024
  git-commit: 87d898ad4
  cantera-version: 3.1.0a4
  generator: Cantera SolutionArray
  reactants:
    type: inlet
    size: 1
    points: 1
    mass-flux: 2.709217648400740
    temperature: 300.0
    pressure: 1.01325e+05
    mass-fractions:
      H2: 8.338727723581104e-03
      O2: 0.1654406062122694
      AR: 0.8262206660641496
  flame:
    type: axisymmetric-flow
    size: 20
    basis: mass
    points: 20
    energy-enabled: true
    phase:
      name: ohmech
      source: /Volumes/Data/work/GitHub/cantera/test/python/../../build/python/cantera/data/h2o2.yaml
    radiation-enabled: false
    species-enabled: true
    tolerances:
      transient-abstol: 1.0e-11
      steady-abstol: 1.0e-09
      transient-reltol: 1.0e-04
      steady-reltol: 1.0e-04
    transport-model: mixture-averaged
    Soret-enabled: false
    flux-gradient-basis: 1
    refine-criteria:
      ratio: 5.0
      slope: 0.8
      curve: 1.0
      prune: 0.2
      grid-min: 1.0e-10
      max-points: 1000
    components: [grid, velocity, spread_rate, lambda, T, D, H2, H, O, O2, OH, H2O, HO2,
    H2O2, AR, N2]
    grid: [0.0, 0.01, 0.0203125, 0.02515625, 0.02875, 0.03015625, 0.0309375,
    0.031171875, 0.03125, 0.03140625, 0.031484375, 0.0315625, 0.03171875,
    0.031875, 0.0321875, 0.0325, 0.03375, 0.035, 0.04, 0.05]
    lambda: [-3933.031794147984, -3933.031794147984, -3933.031794147984,
    -3933.031794147984, -3933.031794147984, -3933.031794147984,
    -3933.031794147984, -3933.031794147984, -3933.031794147984,
    -3933.031794147984, -3933.031794147984, -3933.031794147984,
    -3933.031794147984, -3933.031794147984, -3933.031794147984,
    -3933.031794147984, -3933.031794147984, -3933.031794147984,
    -3933.031794147984, -3933.031794147984]
    spread_rate: [2.195507286868176e-20, 14.51722343698459, 29.48811000955164,
    36.51988188853007, 41.73664845563419, 43.77685091273374,
    44.91384471788903, 45.27340040953393, 45.41578426558767,
    45.80104196650910, 46.05703977533385, 46.36593136241802,
    47.06428364748030, 47.76270236099705, 49.15657145226285,
    50.54501272403288, 56.11685417990889, 61.69162860926554,
    83.70320438491522, 83.70320438491522]
    velocity: [1.999999999998790, 1.854827765341866, 1.401022707938707,
    1.081294119788945, 0.7999951866950966, 0.6794271565417825,
    0.6105613842818222, 0.5998153941960703, 0.6102814971867941,
    0.7006948286465666, 0.8310351307898834, 1.069950110824169,
    1.760946682043529, 2.263305098387171, 2.532867685591085,
    2.602054114337823, 2.527435208512887, 2.394167502385834, 1.6740640876979,
    3.406594406359178e-37]
    T: [300.0, 300.0000000006364, 300.0000003634784, 300.0000535458819,
    300.0053670437679, 300.1294232841556, 301.7084741967649,
    309.2726321919635, 320.0041220934847, 381.3297912376120,
    460.9446778944669, 603.3683494450601, 1015.533184385543,
    1320.396804449897, 1502.143068341387, 1565.332091098152,
    1604.032638002299, 1613.862043613176, 1619.580422859907,
    1619.580422859907]
```

**Review by ischoegl — COMMENTED** [review]
Here's one thing we could use to check the correct order:
```python
def check_component_order(fname: str, group: str):
    with fname.open("r", encoding="utf-8") as fid:
        reader = yaml.YAML(typ="safe")
        contents = reader.load(fid)
    components = []
    index = -1
    for key, value in contents[group]["flame"].items():
        if components:
            index += 1
            field = components.pop()
            msg = f"Found mismatch of components at index {index}: expected {field!r}"
            assert key == field, msg
        if key == "components":
            components = value[::-1]
```

I added this to `test_onedim.py` as well as a couple of strategic edits (I think there's a total of 6 instances):
```
    def test_restore_yaml(self):
        filename, group = self.run_save_restore("yaml")
        check_component_order(filename, group)
```

and it shows the following failures:
```none
FAILED test/python/test_onedim.py::TestStagnationFlame::test_restore_yaml - AssertionError: Found mismatch of components at index 1: expected 'velocity'
FAILED test/python/test_onedim.py::TestImpingingJet::test_restore_yaml - AssertionError: Found mismatch of components at index 1: expected 'velocity'
FAILED test/python/test_onedim.py::TestTwinFlame::test_save_restore_yaml - AssertionError: Found mismatch of components at index 1: expected 'velocity'
```
Fwiw, the commit is [here](https://github.com/ischoegl/cantera/commit/d6e0a50526903b618ee6f7ddd821fb274465dbc4)

**Review by ischoegl — APPROVED** [review]
:tada: … thanks, @speth. This looks good to me!

**Pull request merged**